odoo/addons/base/models/qweb.py

```diff
@@ -4,150 +4,82 @@
 import os.path
 import re
 import traceback
-
-from collections import OrderedDict
-from collections.abc import Sized, Mapping
-from functools import reduce
-from itertools import tee, count
-from textwrap import dedent
-
-import itertools
-from lxml import etree, html
+import builtins
+import types
+import token
+import tokenize
+import io
+from itertools import count, chain
+from textwrap import dedent, indent as _indent
+from time import time
+
+from lxml import etree
 from psycopg2.extensions import TransactionRollbackError
 import werkzeug
 from werkzeug.utils import escape as _escape
 
-from odoo.tools import pycompat, freehash
-from odoo.tools.safe_eval import check_values
-
-import builtins
-builtin_defaults = {name: getattr(builtins, name) for name in dir(builtins)}
-
-try:
-    import astor
-except ImportError:
-    astor = None
-
-from odoo.tools.parse_version import parse_version
-
-unsafe_eval = eval
+from odoo.tools import pycompat, freehash, parse_version
 
 _logger = logging.getLogger(__name__)
 
 ####################################
 ###          qweb tools          ###
 ####################################
 
+# to remove after v15
+class Qweb_deprecated_foreach():
```

Comment:
What is this?  Please write some documentation.

Reply:
Just a object to have error when use old value of t-foreach. To help user to found errors. After will remove it